sklearn/ensemble/iforest.py

```diff
@@ -0,0 +1,266 @@
+        of the leaf containing this observation, which is equivalent to
+        the number of splitting required to isolate this point. In case of
+        several observations n_left in the leaf, the average length path of
+        a n_left samples isolation tree is added.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix of shape (n_samples, n_features)
+            The input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csr_matrix``.
+
+        Returns
+        -------
+        scores : array of shape (n_samples,)
+            The anomaly score of the input samples.
+            The lower, the more normal.
+        """
+        # code structure from ForestClassifier/predict_proba
+        # Check data
+        X = self.estimators_[0]._validate_X_predict(X, check_input=True)
+        n_samples = X.shape[0]
+
+        # Assign chunk of trees to jobs
+        n_jobs, n_trees, starts = _partition_estimators(self.n_estimators,
+                                                        self.n_jobs)
+
+        # Parallel loop
+        results = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,
+                           backend="threading")(
+                               delayed(_parallel_helper)(tree.tree_,
+                                                         'apply_depth', X)
+                               for tree in self.estimators_)
+
+        # We had to add to the depth (results) the average depth wrt to
+        # the number of samples remaining in the leaf: First find this number:
+        n_samples_leaf = np.zeros((self.n_estimators, n_samples))
+        estimator_nb = -1
+        for tree in self.estimators_:
+            estimator_nb += 1
+            leaves_index = tree.apply(X)
+            n_nodes = tree.tree_.n_node_samples
+            n_samples_leaf[estimator_nb, :] = n_nodes[leaves_index]
+
+        # Now, the depth if the tree was fully grown: (add an average term)
+        depth_if_full = np.array(results) + _cost(np.array(n_samples_leaf))
+        scores = np.zeros(n_samples)
+        depth = np.mean(depth_if_full, axis=0)
+
+        scores = np.power(2 * np.ones(n_samples), -depth /
+                          _cost(self.max_samples)[0, 0])
+
+        return scores
+
+    def decision_function(self, X):
+        """Average of the decision functions of the base classifiers.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            The training input samples. Sparse matrices are accepted only if
+            they are supported by the base estimator.
+
+        Returns
+        -------
+        score : array, shape (n_samples,)
+            The decision function of the input samples.
+
+        """
+        # minus as bigger is better (here less abnormal):
+        return - self.predict(X)
+
+
+def _cost(n_samples):
```

Comment:
It's not clear that n_samples is an array. I would have assume that it is a integer from the name and conventions.